internal/pagination/pagination2.go

```diff
@@ -0,0 +1,189 @@
+// Copyright (c) HashiCorp, Inc.
+// SPDX-License-Identifier: BUSL-1.1
+
+package pagination
+
+import (
+	"context"
+	"time"
+
+	"github.com/hashicorp/boundary/internal/boundary"
+	"github.com/hashicorp/boundary/internal/errors"
+	"github.com/hashicorp/boundary/internal/refreshtoken"
+)
+
+// ListResponse2 represents the response from the paginated list operation.
+type ListResponse2[T boundary.Resource] struct {
+	Items              []T
+	CompleteListing    bool
+	RefreshToken       *refreshtoken.Token
+	DeletedIds         []string
+	EstimatedItemCount int
+}
+
+// ListFilterFunc is a callback passed down from the application service layer
+// used to filter out resources that don't match any user-supplied filter.
+type ListFilterFunc[T boundary.Resource] func(ctx context.Context, item T) (bool, error)
+
+// ListItemsFunc2 is used to encapsulate the creation of repository specific options
+// and listing of items.
+type ListItemsFunc2[T boundary.Resource] func(ctx context.Context, prevPageLastItem T, limit int) ([]T, error)
+
+// EstimatedCountFunc is used to estimate the total number of items
+// available in the table given any filter provided.
+type EstimatedCountFunc func(ctx context.Context) (int, error)
+
+// ListDeletedIDsFunc is used to list the IDs of the resources deleted since
+// the given timestamp. It returns a slice of IDs and the timestamp of the
+// instant in which the slice was created.
+type ListDeletedIDsFunc func(ctx context.Context, since time.Time) ([]string, time.Time, error)
+
+// List lists items according to the page size,
+// filtering out entries that do not
+// pass the filter item fn. It returns a new refresh token
+// based on the grants hash and the returned items.
```

Comment:
I think this could include additional details to explain more of what this function does, for example:

```go
// List returns a ListResponse. The response will contain at most a
// number of items equal to the pageSize. Items are fetched using the
// listItemsFn and then items are checked using the filterItemFn
// to determine if they should be included in the response.
// The response includes a new refresh token based on the grants and items.
// The estimatedCountFn is used to provide an estimated total number of
// items that can be returned by making additional requests using the provided
// refresh token.